pilot/pkg/proxy/envoy/v2/datastructures.go

```diff
@@ -0,0 +1,139 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+package v2
+
+import (
+	"sync"
+
+	xdsapi "github.com/envoyproxy/go-control-plane/api"
+	"istio.io/istio/pilot/pkg/model"
+)
+
+// Endpoints (react to updates in registry endpoints) (notify clusters)
+// 1. Endpoints can be shared across multiple clusters.
+// 2. The membership of an endpoint in a cluster is based on the set of labels
+//    associated with the endpoint and the set of label selectors associated with the cluster.
+// 3. When a new endpoint is added, the label selectors associated with the clusters have
+//    to be evaluated to determine the clusters to which the endpoint belongs to.
+// 4. When an endpoint is deleted, it needs to be removed from all associated clusters.
+// 5. When one or more labels are added to an endpoint, the label selectors from clusters
+//    that "do not have this endpoint" need to be reevaluated to see if the newly added labels
+//    would cause this endpoint to be added to these clusters.
+// 6. When one or more labels are "removed" from an endpoint, the clusters currently associated
+//    with the endpoint need to be reevaluated to see if the endpoint still qualifies to be a member
+//    of these clusters.
+
+// Clusters (react to updates in services, dest rules, ext services)
+// 1. Every service (registry defined, external) in the system has a corresponding cluster
+// 2. Every subset expressed in a DestinationRule will have an associated cluster.
+// 3. Clusters are added/removed when a service is added/removed or when subsets are
```

Comment:
I already noted the subsets thing.. Yes, if a new service is added on a port that does not exist already, then new listeners are needed.. This LDS push goes to *all* nodes, because in kube, its all to all networking.